In [22]:
# import necessary librarys

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline


In [23]:
data = pd.read_csv('train.csv')

In [24]:
data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [25]:
print(data.columns)

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

In [26]:
data_ID = data['Id']

sale_prices = data['SalePrice']

data.drop(['Id', 'SalePrice'], axis = 1, inplace = True)

In [36]:
data['Utilities'].value_counts()

AllPub    1459
NoSeWa       1
Name: Utilities, dtype: int64

In [ ]:
data.drop('Utilities', axis = 1, inplace = True)

In [37]:
numerical_features = ['LotFrontage','LotArea','MasVnrArea','BsmtFinSF1','BsmtFinSF2',
                     'BsmtUnfSF','TotalBsmtSF','1stFlrSF','2ndFlrSF','LowQualFinSF',
                      'GrLivArea','GarageArea','WoodDeckSF','OpenPorchSF','EnclosedPorch',
                     '3SsnPorch','ScreenPorch','PoolArea','MiscVal']

**Featureset Exploration**
* MSSubClass: 1 to 14 in numerical format, but it should be considered as categorical variable.
* MSZoning: RL, RM, FV, RH and C.
* LotFrontage: continuous.
* LotArea: continuous.
* Street: Pave and Grvl.
* Alley: Pave and Grvl.
* LotShape: Reg, IR1, IR2 and IR3.
* LandContour: Lvl, Bnk, HLS and Low.
* Utilities: AllPub(1459) and NoSeWa(1)
* LotConfig: Inside, Corner, CulDSac, FR2 and FR3.
* LandSlope: Gtl, Mod and Sev.
* Neighborhood: NAmes, CollgCr, OldTown, etc. 
* Condition1: Norm, Feedr, Artery, etc.
* Condition2: Norm, Feedr, Artery, etc.
* BldgType: 1Fam, TwnhsE, Duplex, Twnhs and 2fmCon.
* HouseStyle: 1Story, 2Story, 1.5Fin, etc.
* OverallQual: 1 to 10 in numerical format, but it should be considered as categorical variable.
* OverallCond: 1 to 9 in numerical format, but it should be considered as categorical variable.
* YearBuilt: 
* YearRemodAdd: 
* RoofStyle: Gable, Hip, Flat, Gambrel, Mansard and Shed.
* RoofMatl: CompShg, Tar&Grv, WdShngl, etc.
* Exterior1st: VinylSd, HdBoard, MetalSd, etc.
* Exterior2nd: VinylSd, HdBoard, MetalSd, etc.
* MasVnrType: BrkFace, Stone and BrkCmn.
* MasVnrArea: continuous.
* ExterQual: TA, Gd, Ex and Fa.
* ExterCond: TA, Gd, Ex, Fa and Po.
* Foundation: PConc, CBlock, BrkTil, Slab, Stone and Wood.
* BsmtQual: TA, Gd, Ex and Fa.
* BsmtCond: TA, Gd, Fa and Po.
* BsmtExposure: No, Av, Gd and Mn.
* BsmtFinType1: Unf, GLQ, ALQ, BLQ, Rec and LwQ.
* BsmtFinSF1: continuous.
* BsmtFinType2: Unf, GLQ, ALQ, BLQ, Rec and LwQ.
* BsmtFinSF2: continuous.
* BsmtUnfSF: continuous.
* TotalBsmtSF: continuous.
* Heating: GasA, GasW, Grav, Wall, OthW and Floor.
* HeatingQC: Ex, TA, Gd, Fa and Po.
* CentralAir: Y and N.
* Electrical: SBrkr, FuseA, FuseF, FuseP and Mix.
* 1stFlrSF: continuous.
* 2ndFlrSF: continuous.
* LowQualFinSF: continuous.
* GrLivArea: continuous.
* BsmtFullBath: 0, 1, 2 and 3.
* BsmtHalfBath: 0, 1 and 2.
* FullBath: 0, 1, 2 and 3.
* HalfBath: 0, 1 and 2.
* BedroomAbvGr: 0 to 8.
* KitchenAbvGr: 0, 1, 2 and 3.
* KitchenQual: TA, Gd, Ex and Fa.
* TotRmsAbvGrd: 2 to 14.
* Functional: Typ, Min1, Min2, Mod, Maj1, Maj2 and Sev.
* Fireplaces: 0, 1, 2 and 3.
* FireplaceQu: Ex, TA, Gd, Fa and Po.
* GarageType: Attchd, Detchd, BuiltIn, Basment, CarPort and 2Types.
* GarageYrBlt: 
* GarageFinish: Unf, RFn and Fin.
* GarageCars: 0 to 4.
* GarageArea: continuous.
* GarageQual: Ex, TA, Gd, Fa and Po.
* GarageCond: Ex, TA, Gd, Fa and Po.
* PavedDrive: Y, N and P.
* WoodDeckSF: continuous.
* OpenPorchSF: continuous.
* EnclosedPorch: continuous.
* 3SsnPorch: continuous.
* ScreenPorch: continuous.
* PoolArea: continuous.
* PoolQC: Gd, Ex and Fa.
* Fence: MnPrv, GdPrv, GdWo and MnWw.
* MiscFeature: Shed, Gar2, Othr and TenC.
* MiscVal: continuous.
* MoSold: 1 to 12 month.
* YrSold: 
* SaleType: WD, New, COD, etc.
* SaleCondition: Normal, Partial, Abnorml, Family, Alloca and AdjLand.




In [27]:
data['SaleCondition'].value_counts()

Normal     1198
Partial     125
Abnorml     101
Family       20
Alloca       12
AdjLand       4
Name: SaleCondition, dtype: int64

In [28]:
data.describe()

MSSubClass  LotFrontage        LotArea  OverallQual  OverallCond  \
count  1460.000000  1201.000000    1460.000000  1460.000000  1460.000000   
mean     56.897260    70.049958   10516.828082     6.099315     5.575342   
std      42.300571    24.284752    9981.264932     1.382997     1.112799   
min      20.000000    21.000000    1300.000000     1.000000     1.000000   
25%      20.000000    59.000000    7553.500000     5.000000     5.000000   
50%      50.000000    69.000000    9478.500000     6.000000     5.000000   
75%      70.000000    80.000000   11601.500000     7.000000     6.000000   
max     190.000000   313.000000  215245.000000    10.000000     9.000000   

         YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1   BsmtFinSF2  \
count  1460.000000   1460.000000  1452.000000  1460.000000  1460.000000   
mean   1971.267808   1984.865753   103.685262   443.639726    46.549315   
std      30.202904     20.645407   181.066207   456.098091   161.319273   
min    1872.000000   1950.000000     0.000000     0.000000     0.000000   
25%    1954.000000   1967.000000     0.000000     0.000000     0.000000   
50%    1973.000000   1994.000000     0.000000   383.500000     0.000000   
75%    2000.000000   2004.000000   166.000000   712.250000     0.000000   
max    2010.000000   2010.000000  1600.000000  5644.000000  1474.000000   

          ...        GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch  \
count     ...       1460.000000  1460.000000  1460.000000    1460.000000   
mean      ...        472.980137    94.244521    46.660274      21.954110   
std       ...        213.804841   125.338794    66.256028      61.119149   
min       ...          0.000000     0.000000     0.000000       0.000000   
25%       ...        334.500000     0.000000     0.000000       0.000000   
50%       ...        480.000000     0.000000    25.000000       0.000000   
75%       ...        576.000000   168.000000    68.000000       0.000000   
max       ...       1418.000000   857.000000   547.000000     552.000000   

         3SsnPorch  ScreenPorch     PoolArea       MiscVal       MoSold  \
count  1460.000000  1460.000000  1460.000000   1460.000000  1460.000000   
mean      3.409589    15.060959     2.758904     43.489041     6.321918   
std      29.317331    55.757415    40.177307    496.123024     2.703626   
min       0.000000     0.000000     0.000000      0.000000     1.000000   
25%       0.000000     0.000000     0.000000      0.000000     5.000000   
50%       0.000000     0.000000     0.000000      0.000000     6.000000   
75%       0.000000     0.000000     0.000000      0.000000     8.000000   
max     508.000000   480.000000   738.000000  15500.000000    12.000000   

            YrSold  
count  1460.000000  
mean   2007.815753  
std       1.328095  
min    2006.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2010.000000  

[8 rows x 36 columns]

In [29]:
null_sum = data.isnull().sum()
null_sum[null_sum>0].sort_values(ascending=False)
#aa[aa>0].sort_values(ascending=False)
#print(data.isnull().sum())

PoolQC          1453
MiscFeature     1406
Alley           1369
Fence           1179
FireplaceQu      690
LotFrontage      259
GarageYrBlt       81
GarageType        81
GarageFinish      81
GarageQual        81
GarageCond        81
BsmtFinType2      38
BsmtExposure      38
BsmtFinType1      37
BsmtCond          37
BsmtQual          37
MasVnrArea         8
MasVnrType         8
Electrical         1
dtype: int64

In [38]:
data['MasVnrArea'].fillna(0, inplace=True)
data['Electrical'].fillna(data['Electrical'].mode()[0], inplace=True)

In [39]:
cols_fill_with_none = ['PoolQC' , 'MiscFeature', 'Alley', 'Fence', 'FireplaceQu', 'GarageQual', 
         'GarageCond', 'GarageFinish', 'GarageYrBlt', 'GarageType', 'BsmtExposure', 
         'BsmtCond', 'BsmtQual', 'BsmtFinType2', 'BsmtFinType1', 'MasVnrType']
for col in cols_fill_with_none:
    data[col].fillna("None", inplace=True)

In [40]:
"""
LotFrontage: Since the area of each street connected to the house property
most likely have a similar area to other houses in its neighborhood,
we can fill in missing values by the median LotFrontage of the neighborhood.
"""
data["LotFrontage"] = data.groupby("Neighborhood")["LotFrontage"].transform(lambda x: x.fillna(x.median()))

In [41]:
#Convert some numerical features into categorical features.
#It's better to use LabelEncoder and get_dummies for these features.

NumStr = ['MSSubClass','BsmtFullBath','BsmtHalfBath','HalfBath','BedroomAbvGr','KitchenAbvGr',
          'MoSold','YrSold','YearBuilt','YearRemodAdd','LowQualFinSF','GarageYrBlt']
for col in NumStr:
    data[col] = data[col].astype(str)

In [42]:
from sklearn.preprocessing import LabelEncoder
cols = ('FireplaceQu', 'BsmtQual', 'BsmtCond', 'GarageQual', 'GarageCond', 
        'ExterQual', 'ExterCond','HeatingQC', 'PoolQC', 'KitchenQual', 'BsmtFinType1', 
        'BsmtFinType2', 'Functional', 'Fence', 'BsmtExposure', 'GarageFinish', 'LandSlope',
        'LotShape', 'PavedDrive', 'Street', 'Alley', 'CentralAir', 'MSSubClass', 'OverallCond', 
        'YrSold', 'MoSold')
# process columns, apply LabelEncoder to categorical features
for c in cols:
    lbl = LabelEncoder() 
    lbl.fit(list(data[c].values)) 
    data[c] = lbl.transform(list(data[c].values))

# shape        
print('Shape all_data: {}'.format(data.shape))

Shape all_data: (1460, 79)


In [43]:
from sklearn.preprocessing import MinMaxScaler
numeric_feats = data.dtypes[data.dtypes != "object"].index
print (numeric_feats)

Index(['MSSubClass', 'LotFrontage', 'LotArea', 'Street', 'Alley', 'LotShape',
       'LandSlope', 'OverallQual', 'OverallCond', 'MasVnrArea', 'ExterQual',
       'ExterCond', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1',
       'BsmtFinSF1', 'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF',
       'HeatingQC', 'CentralAir', '1stFlrSF', '2ndFlrSF', 'GrLivArea',
       'FullBath', 'KitchenQual', 'TotRmsAbvGrd', 'Functional', 'Fireplaces',
       'FireplaceQu', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive', 'WoodDeckSF', 'OpenPorchSF',
       'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'PoolQC',
       'Fence', 'MiscVal', 'MoSold', 'YrSold'],
      dtype='object')


In [ ]:
final_data = pd.get_dummies(data)
display(final_data.head(n=5))
print(final_data.columns)

In [ ]:
plt.scatter(x=data.GrLivArea, y=data.SalePrice)
plt.xlabel("GrLivArea", fontsize=13)
plt.ylabel("SalePrice", fontsize=13)
plt.ylim(0,800000)

In [ ]:
#Deleting outliers
good_data = data.drop(data[(data['GrLivArea']>4000) & (data['SalePrice']<300000)].index)

In [ ]:
# most correlated features
corrmat = data.corr()
top_corr_features = corrmat.index[abs(corrmat["SalePrice"])>0.5]
plt.figure(figsize=(10,10))
sns.heatmap(data[top_corr_features].corr(),annot=True)

In [ ]:
print(sns.distplot(data['SalePrice']))

In [ ]:
sns.distplot(np.log(data['SalePrice']))

In [ ]:
good_data['SalePrice'] = np.log(good_data['SalePrice'])